### PR TITLE
Include floats in the number notation mechanism

### PR DESCRIPTION
**Pull request by proux01**
Number Notation can now be used to print float numeric values.
See theories/Floats/PrimFloat.v for an example.
This incidentally removes the float syntax plugin.

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Review by herbelin — APPROVED** [review]
I did not enter into all details but this looks very nice and carefully designed.

The PR title is a bit misguiding though. Rather than removing an ad hoc notation plugin, it is about providing user-definable notations for floating-point numbers.

A couple of examples in the PR header would help to give an idea of how it works in practice (and of course, examples in the test-suite would eventually be useful).

**Comment by proux01** [comment]
Thanks for the review, I updated the title and description and added a test in the test-suite (for the feature that is not used in the stdlib).

**Comment by Zimmi48** [comment]
> I updated the title and description

It looks like you didn't update the title, and you updated the description in the reverse direction compared to what Hugo had suggested :confused:

File: theories/Floats/PrimFloat.v [hunk 1]
```diff
@@ -25,15 +25,20 @@ Register float_class as kernel.ind_f_class.
 (** ** The main type *)
 (** [float]: primitive type for Binary64 floating-point numbers. *)
 Primitive float := #float64_type.
+Register float as num.float.type.
+
+Record float_wrapper := wrap_float { float_wrap : float }.
```
Comment (JasonGross) [inline theories/Floats/PrimFloat.v:30]: Question: Will all the code still work if this is a record with `Primitive Projections` set?

File: theories/Floats/PrimFloat.v [hunk 2]
```diff
@@ -25,15 +25,20 @@ Register float_class as kernel.ind_f_class.
...
+Register wrap_float as num.float.wrap_float.
+Definition printer (x : float_wrapper) : option Number.number := None.
```
Comment (JasonGross) [inline theories/Floats/PrimFloat.v:32]: Why no printer here?  Is it because primitive floats already have a printer?

File: test-suite/output/NumberNotations.v [hunk 3]
```diff
@@ -978,3 +978,23 @@ Fail Check 3 : Fin.t 3.
...
+
+Definition parser (x : float) : float := x.
+
+Number Notation float parser printer : float_scope.
+
+Check 12%float.
+Check (-12)%float.
+Check infinity.
+Check neg_infinity.
+Check nan.
```
Comment (JasonGross) [inline test-suite/output/NumberNotations.v:998]: Where does parsing/printing of these happen?  Is it just defined constants?  What if you use `Compute` instead of `Check`?

**Review by JasonGross — CHANGES_REQUESTED** [review]
Looks good (though I haven't looked over it in detail)!  I'm glad to have the float syntax plugin be removed, though it looks like this PR doesn't actually remove the files even though it removes use of the ML module?  I take it that's an oversight?

File: theories/Floats/PrimFloat.v (hunk 1 above)
Comment (proux01) [reply theories/Floats/PrimFloat.v:30]: I would have expected this to fail, but it seems to work.

File: theories/Floats/PrimFloat.v (hunk 2 above)
Comment (proux01) [reply theories/Floats/PrimFloat.v:32]: `Number Notation` can only parse/print numbers, however floats have some non numeric values (infinities and NaN), for parsing, only numeric values are parsed but for printing, we could not handle the result of the printing function if it returns one of those non numeric values, so we simply forbid printers to return floats. The printer remains hardcoded there https://github.com/coq/coq/blob/master/interp/constrextern.ml#L920

Maybe another solution could be to allow printer returning float and just make them fail when they return non numeric values.

File: test-suite/output/NumberNotations.v (hunk 3 above)
Comment (proux01) [reply test-suite/output/NumberNotations.v:998]: These are just constants https://github.com/coq/coq/blob/master/theories/Floats/PrimFloat.v#L107 so no specific parsing needed. Printing is hardcoded https://github.com/coq/coq/blob/master/interp/constrextern.ml#L920

`Compute` wouldn't change anything since there is nothing left to compute.

File: theories/Floats/PrimFloat.v (hunk 2 above)
Comment (proux01) [reply theories/Floats/PrimFloat.v]: Eventually, I implemented this last solution, this is not more complex but probably more principled: printers can return float but are ignore when returning NaN or infinities.

**Comment by proux01** [comment]
@JasonGross your comments should be adressed

**Comment by Zimmi48** [comment]
ping @JasonGross

**Comment by JasonGross** [comment]
Sorry, I'm currently quite busy, but hope to find time within the next couple of days.  If I haven't commented again by Monday, it's probably because it fell off my stack again (I'm in the process of moving homes), so feel free to ping me again then.

**Comment by proux01** [comment]
No worries, we are not in a hurry, there is plenty of time for 8.15.

**Review by JasonGross — APPROVED** [review]
Changes look good, I've skimmed through the code again and things look good at a glance to me

**Comment by proux01** [comment]
Thanks @JasonGross, would you be the assignee?
I'm not sure we can still target 8.14 (getting this only in 8.15 would be fine by me).

**Comment by JasonGross** [comment]
@coqbot merge now

**Comment by JasonGross** [comment]
Sorry for the delay in getting around to this

**Comment by proux01** [comment]
@JasonGross no worries, thanks for your, as always, careful review.

**Comment by proux01** [comment]
Thanks for the suggestion, we should find a way to keep the warning when parsing inexact floats (c.f. #11859 ) and we could then go with that solution.

**Comment by JasonGross** [comment]
Perhaps if we merge @herbelin 's reduction effects plugin into Coq proper and add support for emitting warnings and not just messages, that would let us emit the warning messages during parsing?

**Comment by proux01** [comment]
@JasonGross sounds great, would you have a more precise pointer on Hugo's work?

**Comment by andrew-appel** [comment]
By the way, I have updated float_notations.v in my own (currently private) repo to replace "rebuild_float" with "erase"; if and when you are ready to use this code, just ping me by e-mail, instead of using the one that I posted here.

**Comment by JasonGross** [comment]
@proux01 https://github.com/coq-community/reduction-effects, see in particular https://github.com/coq-community/reduction-effects/blob/master/PrintingEffect.v and https://github.com/coq-community/reduction-effects/blob/master/tests/PrintEffect.v and https://github.com/coq-community/reduction-effects/blob/master/tests/PrintEffect.expected

**Comment by proux01** [comment]
@JasonGross thanks for the pointer, that sounds good indeed. @ppedrot fears that it's "a hack" and "it's not really guaranteed to fire properly" (https://coq.zulipchat.com/#narrow/stream/237977-Coq-users/topic/Triggering.20a.20warning.20during.20Compute), any bad experience on your side? (I would tend to consider that since it's just to display a warning, we don't need it to be absolutely 100% reliable)

**Comment by JasonGross** [comment]
@ppedrot You're worried that adding to `constant_effect_table` will impact reduction performance globally by slowing down https://github.com/coq/coq/blob/bdb72aab60f40172743de2bbfd111c22037652c3/pretyping/reductionops.ml#L47 ?

@ppedrot Can you explain a bit more?  Which part do you consider a hack?  And what's wrong with adding a vernacular (or attribute, or whatever) that lets users declare that certain constants, when reduced, should emit warning messages which are constructed from the arguments passed to the constant?

**Comment by ppedrot** [comment]
> Which part do you consider a hack?

The are several problems.

First thing, this semantics only makes sense for call-by-value reduction, because it will break observational equivalence in the call-by-name equational theory. Here it's more or less fine because we use `cbv` but if ever we want to get slightly more complex we're going to get weird behaviours. But in general, Coq equational theory is call-by-name, so this mechanism is at odds with the very basic semantics of the system. The only reasonable alternative system I could think of that is both compatible with cbn and cbv is that the effectful constant must have an inductive type, and the effect must be performed when it's matched, *not* unfolded.

Second, this is going to trigger warnings whenever somebody starts using the wrappers used for numeral printing in daily code. We can recommend against their use, but I wouldn't be surprised these might pop up in normal use due to tactics.

Third, the implementation relies on the existence of side-effects at the level of OCaml code itself. This means that it's very hard to control that some code using a reduction machine for other purposes is not going to randomly output warnings because it so happens that there is a constant in there that's supposed to raise warning for numeral parsing.

Really, this feature shouldn't be used outside of hackish plugins that want to get things done. But for numeral printing, we need a more explicit, functional API that is not going to unexpectedly interfere with anything else without even respecting the Coq semantics anyways. On Zulip, I advocated for allowing the return values of numeral evaluation to also contain a list of strings as messages, i.e. the good old writer monad. We already do it with an `option` for partial functions, I don't see why this couldn't be extended to arbitrary effects. (We could even do this in a more principled way à la Mtac2 with an opaque datatype hiding the monadic interface to the user, allowing backwards compatibility when a new effect is added).

**Comment by herbelin** [comment]
> First thing, this semantics only makes sense for call-by-value reduction, because it will break observational equivalence in the call-by-name equational theory. Here it's more or less fine because we use cbv but if ever we want to get slightly more complex we're going to get weird behaviours. But in general, Coq equational theory is call-by-name, so this mechanism is at odds with the very basic semantics of the system.

To nitpick: Coq's `cbv` does not implement a call-by-value equational theory (in the sense of Plotkin). It is a call-by-value evaluation order for the call-by-name equational theory. In themselves, cbn and cbv equational semantics are not able to observe the order in which their redexes are contracted. What the unfolding of constants is able to observe is not a difference of equational theory but a difference of evaluation order within Coq call-by-name equational theory.

**Comment by andrew-appel** [comment]
TL;DR:  why do we need this warning at all?
Now I'm confused:  is this intended for numeral parsing or numeral printing?  For printing we don't need it:   for any given float precision (bits in mantissa, bits in exponent) it's always possible to print out a decimal representation in scientific notation that will parse back in as the same number, and, indeed, the solution I propose (written in Coq) does that -- provably (partially) correctly, by validation, as explained in the file I attached.
   In fact, if validation were to fail for some float in my solution, then the float->option(Number.number) translation returns None, which causes Coq's pretty printer to default to the underlying inductive representation.  For me that works, because the "float" in question is a Flocq float, not a Coq primitive float; and Flocq floats are indeed an inductive type that have an ordinary printing representation in Coq.  If we apply my solution to primitive floats, then a simple adjustment to my program would be able to default over to a basic ugly-printable representation via PrimFloat, in the (possibly nonexistent) case where validation fails.

**Comment by proux01** [comment]
You are right, the warning are only needed for parsing (for instance, it is nice to warn the user that parsing 0.1 doesn't give the real number 1/10).
For printing, we can indeed resort to ugly representations in case of difficulty (either via PrimFloat, or simply in hexadecimal format).

**Comment by andrew-appel** [comment]
In my humble opinion, this warning is not needed at all.  There's a well defined total function from scientific-notation (decimal) representations to floats (of a given precision).  It is our responsibility to implement this function accurately, and it is the user's responsibility to understand that not every scientific-decimal-notation number is exactly represented.
   Reductio ad absurdum:  Would you also propose that floating-point subtraction should give a warning if the result is not exact?

**Comment by andrew-appel** [comment]
Um, actually, it's not necessarily a total function;  my own implementation rejects underflows and overflows (instead of mapping them to zeros and infinities).  But that results in an error message (when the function from Number.number->option(float) produces None), not a warning, so my argument still stands.

**Comment by JasonGross** [comment]
> First thing, this semantics only makes sense for call-by-value reduction

This is not an issue, here, at least, as long as we ensure that in numeral parsing we are forced to unfold the function at the right time regardless of strategy.  I claim this is easy for parsing.

> The only reasonable alternative system I could think of that is both compatible with cbn and cbv is that the effectful constant must have an inductive type, and the effect must be performed when it's matched, not unfolded.

I endorse this alternative, it seems better semantically.

> Second, this is going to trigger warnings whenever somebody starts using the wrappers used for numeral printing in daily code

This is easy to avoid.  We only need the warning at the end of parsing (don't need it at all in printing), so we can have the core of the functionality in a function that doesn't use this constant and returns a pair of the parsed value and the error info.  Then we can provide a standard function which is just the fst of that function, and another function-with-warning that first emits the warning and then returns the value.  This later function should be used only by the number notation mechanism, and nowhere else.

> We can recommend against their use, but I wouldn't be surprised these might pop up in normal use due to tactics.

The function with warning should not show up even indirectly in any term outside of the file in which the number notation is added.  We can add a comment above it to not use it in proofs.  It will not show up in parsing because parsing fully evaluates the term. So I'm not sure how tactics are going to get their hands on it?

> Third, the implementation relies on the existence of side-effects at the level of OCaml code itself. This means that it's very hard to control that some code using a reduction machine for other purposes is not going to randomly output warnings because it so happens that there is a constant in there that's supposed to raise warning for numeral parsing.

The constant should not show up anywhere except when it's being defined and when parsing is happening.  So I guess if `Definition` runs `hnf` and that results in a warning, that could happen, but once the constant is defined, it's not clear to me how other bits of code are going to get their hands on it, except by deliberate user choice.

> Really, this feature shouldn't be used outside of hackish plugins that want to get things done. But for numeral printing, we need a more explicit, functional API that is not going to unexpectedly interfere with anything else without even respecting the Coq semantics anyways. On Zulip, I advocated for allowing the return values of numeral evaluation to also contain a list of strings as messages, i.e. the good old writer monad. We already do it with an option for partial functions, I don't see why this couldn't be extended to arbitrary effects. (We could even do this in a more principled way à la Mtac2 with an opaque datatype hiding the monadic interface to the user, allowing backwards compatibility when a new effect is added).

This also seems fine to me, to bake in the warnings at the number notation mechanism rather than via reduction, but I'm still not convinced this is a big issue for parsing numerals.
